### DISCUSSION TITLE
Any Class Decorators and Generic Arguments

### DISCUSSION THREAD

**Discussion by hmc-cs-mdrissi** [microsoft/pyright#2929] [Q&A]
I'm unsure if this is a bug or as expected. How should Any be handled for generic classes. In this example,

```python
from typing import Any, Generic, TypeVar


def identity(t: Any) -> Any:
    return t


T = TypeVar("T")


@identity
class Foo(Generic[T]):
    ...


@identity
class SubFoo(Foo[T]):
    def bar(self) -> T:
        ...
```

the class type becomes Any. Which leads to type errors like this,

`TypeVar "T" appears only once in generic function signaturePylancereportInvalidTypeVarUse`

I encountered this when working in mypy's codebase. typeshed for mypy_extensions.pyi mark a class decorator like this producing a lot of typevar errors. Specifically this [line](https://github.com/microsoft/pyright/blob/main/packages/pyright-internal/typeshed-fallback/stubs/mypy-extensions/mypy_extensions.pyi) and then running pyright over https://github.com/python/mypy/blob/master/mypy/type_visitor.py.

Maybe this issue is mostly a typeshed issue of using Any for that function just unsure whether Any should not allow type arguments. A simpler example,

```python
class Bar(Any):
   ...

Bar[int] # is this legal or not?
```

**Comment by erictraut** [comment]
Type checking falls apart pretty quickly when an `Any` type appears within a class hierarchy. Is there a compelling use case for why it's being used in this manner here?

**Reply by hmc-cs-mdrissi** [reply]
I'm not really sure as I don't know why mypy_extensions.pyi used Any for class decorator. I've gone ahead and opened a typeshed [issue](https://github.com/python/typeshed/issues/7016) here to see if that stub is needed today or if it's just a stale stub (comment sounds like an old mypy issue hopefully fixed already).
